core/trino-main/src/main/java/io/trino/execution/SqlTaskManager.java

```diff
@@ -528,22 +533,38 @@ private TaskInfo doUpdateTask(
             }
         }
 
-        fragment.map(PlanFragment::getActiveCatalogs)
-                .ifPresent(activeCatalogs -> {
-                    Set<CatalogHandle> catalogHandles = activeCatalogs.stream()
-                            .map(CatalogProperties::getCatalogHandle)
-                            .collect(toImmutableSet());
-                    if (sqlTask.setCatalogs(catalogHandles)) {
-                        ReentrantReadWriteLock.ReadLock catalogInitLock = catalogsLock.readLock();
-                        catalogInitLock.lock();
-                        try {
-                            connectorServicesProvider.ensureCatalogsLoaded(session, activeCatalogs);
-                        }
-                        finally {
-                            catalogInitLock.unlock();
-                        }
+        Optional<List<CatalogProperties>> activeCatalogsOpt = fragment.map(PlanFragment::getActiveCatalogs);
+        if (activeCatalogsOpt.isPresent()) {
+            List<CatalogProperties> activeCatalogs = activeCatalogsOpt.get();
+            Set<CatalogHandle> catalogHandles = activeCatalogs.stream()
+                    .map(CatalogProperties::getCatalogHandle)
+                    .collect(toImmutableSet());
+            if (sqlTask.setCatalogs(catalogHandles)) {
+                ListenableFuture<Void> catalogLoading = Futures.submit(() -> {
+                    ReentrantReadWriteLock.ReadLock catalogInitLock = catalogsLock.readLock();
+                    catalogInitLock.lock();
+                    try {
+                        connectorServicesProvider.ensureCatalogsLoaded(session, activeCatalogs);
```

Comment:
looks like `ensureCatalogsLoaded` could return `ListenableFuture`

Reply:
Wasn't sure how to ensure that the lock is acquired for the execution of the Future only when its needed

Reply:
There is a race condition with pruneCatalog, where a catalog can be pruned immediately after creation, but before use. This lock guards against that

Reply:
add code comment